src/Streamly/Internal/FileSystem/Event/Windows.hs

```diff
@@ -0,0 +1,172 @@
+        return (h, dir)
+
+
+readDirectoryChanges :: Handle -> Bool -> FileNotificationFlag -> IO (Array Word8)
+readDirectoryChanges h wst mask = do
+  let maxBuf = 8*1024
+  allocaBytes maxBuf $ \buffer -> do
+    alloca $ \bret -> do
+       readDirectoryChangesW h buffer (toEnum maxBuf) wst mask bret
+       A.fromStream $ A.toStream $ A.fromPtr maxBuf buffer    
+
+closePathHandleStream :: SerialT IO (Handle, FilePath) -> IO ()
+closePathHandleStream = S.mapM_ ( \(h, _) -> closeHandle h) 
+                      
+
+type FileAction = DWORD
+
+fILE_ACTION_ADDED             :: FileAction
+fILE_ACTION_ADDED             =  1
+fILE_ACTION_REMOVED           :: FileAction
+fILE_ACTION_REMOVED           =  2
+fILE_ACTION_MODIFIED          :: FileAction
+fILE_ACTION_MODIFIED          =  3
+fILE_ACTION_RENAMED_OLD_NAME  :: FileAction
+fILE_ACTION_RENAMED_OLD_NAME  =  4
+fILE_ACTION_RENAMED_NEW_NAME  :: FileAction
+fILE_ACTION_RENAMED_NEW_NAME  =  5
+
+
+-- This is a bit overkill for now, I'll only use nullFunPtr anyway,
+-- but who knows, maybe someday I'll want asynchronous callbacks on the OS level.
+type LPOVERLAPPED_COMPLETION_ROUTINE = FunPtr ((DWORD, DWORD, LPOVERLAPPED) -> IO ())
+
+data FILE_NOTIFY_INFORMATION = FILE_NOTIFY_INFORMATION
+    { fniNextEntryOffset, fniAction :: DWORD
+    , fniFileName :: String
+    } deriving Show
+
+-- instance Storable FILE_NOTIFY_INFORMATION where
+-- ... well, we can't write an instance since the struct is not of fix size,
+-- so we'll have to do it the hard way, and not get anything for free. Sigh.
+
+-- sizeOfFNI :: FILE_NOTIFY_INFORMATION -> Int
+-- sizeOfFNI fni =  (#size FILE_NOTIFY_INFORMATION) + (#size WCHAR) * (length (fniFileName fni) - 1)
+
+peekFNI :: Ptr FILE_NOTIFY_INFORMATION -> IO FILE_NOTIFY_INFORMATION
+peekFNI buf = do
+  neof <- peekByteOff buf 0
+  acti <- peekByteOff buf 4
+  fnle <- peekByteOff buf 8
+  fnam <- peekCWStringLen
+            (buf `plusPtr` 12, -- start of array
+            fromEnum (fnle :: DWORD) `div` 2 ) -- fnle is the length in *bytes*, and a WCHAR is 2 bytes
+  return $ FILE_NOTIFY_INFORMATION neof acti fnam
+
+readDirectoryChangesW :: Handle -> Ptr Word8 -> DWORD -> BOOL -> FileNotificationFlag -> LPDWORD -> IO ()
+readDirectoryChangesW h buf bufSize wst f br =
+  failIfFalse_ "ReadDirectoryChangesW" $ c_ReadDirectoryChangesW h (castPtr buf) bufSize wst f br nullPtr nullFunPtr  
+
+-- The interruptible qualifier will keep threads listening for events from hanging blocking when killed
+
+foreign import ccall safe "windows.h ReadDirectoryChangesW"
+  c_ReadDirectoryChangesW :: Handle -> LPVOID -> DWORD -> BOOL -> DWORD -> LPDWORD -> LPOVERLAPPED -> LPOVERLAPPED_COMPLETION_ROUTINE -> IO BOOL
+
+readOneEvent :: String -> Parser IO Word8 [Event]
+readOneEvent root = do
+  let headerLen = 8*1024
+  arr <- PR.take headerLen (A.writeN headerLen)  
+  fnis <- PR.yieldM $ A.asPtr arr $ \buf -> readChangeEvents buf root
+  return $ fmap maptoEvent fnis  
```

Comment:
We get an Array from `readDirectoryChanges`, we convert that into a stream and then we use the parser to convert it back into an Array. Instead doing this we can directly use the Array returned by readDirectoryChanges and apply `readChangeEvents` on that.